build_tools/cmake/iree_run_module_test.cmake

```
@@ -0,0 +1,371 @@
+#       containing the same.
+#   LABELS: Additional labels to apply to the test. The package path and
+#       "driver=${DRIVER}" are added automatically.
+#   XFAIL: List of platforms (all, x86_64, android-arm64-v8a, riscv64-Linux,
+#       riscv32-Linux) for which the test is expected to fail e.g. due to
+#       issues with the upstream llvm backend. The target will be run, but its
+#       pass/fail status will be inverted.
+#   UNSUPPORTED_PLATFORMS: List of platforms (x86_64, android-arm64-v8a,
+#       riscv64-Linux, riscv32-Linux) not supported by the test target. The
+#       target will be skipped entirely.
+#   DEPS: (Optional) List of targets to build the test artifacts.
+#   TIMEOUT: (optional) Test timeout.
+#
+# Examples:
+#
+# iree_run_module_test(
+#   NAME
+#     iree_run_module_correctness_test
+#   MODULE_SRC
+#     "iree_run_module_bytecode_module_llvm_cpu.vmfb"
+#   DRIVER
+#     "local-sync"
+#   RUNNER_ARGS
+#     "--entry_function=abs"
+#     "--function_input=f32=-10"
+#   EXPECTED_OUTPUT
+#     "f32=10"
+# )
+#
+# iree_run_module_test(
+#   NAME
+#     person_detect_int8_correctness_test
+#   MODULE_SRC
+#     "person_detect_int8.vmfb"
+#   DRIVER
+#     "local-sync"
+#   RUNNER_ARGS
+#     "--entry_function=main"
+#     "--function_input=1x96x96x1xi8=0"
+#   EXPECTED_OUTPUT
+#     "1x2xi8=[72 -72]"
+#   UNSUPPORTED_PLATFORMS
+#     "android-arm64-v8a"
+# )
+#
+# iree_run_module_test(
+#   NAME
+#     mobilenet_v1_fp32_correctness_test
+#   MODULE_SRC
+#     "mobilenet_v1_fp32.vmfb"
+#   DRIVER
+#     "local-sync"
+#   RUNNER_ARGS
+#     "--entry_function=main"
+#     "--function_input=1x224x224x3xf32=0"
+#   EXPECTED_OUTPUT
+#     "mobilenet_v1_fp32_expected_output.txt"
+#   UNSUPPORTED_PLATFORMS
+#     "android-arm64-v8a"
+#     "riscv32-Linux"
+# )
+
+function(iree_run_module_test)
+  if(NOT IREE_BUILD_TESTS)
+    return()
+  endif()
+
+  cmake_parse_arguments(
+    _RULE
+    ""
+    "NAME;MODULE_SRC;DRIVER;EXPECTED_OUTPUT;TIMEOUT"
+    "RUNNER_ARGS;LABELS;XFAIL;UNSUPPORTED_PLATFORMS;DEPS"
+    ${ARGN}
+  )
+
+  iree_get_platform(_PLATFORM)
+  if(_PLATFORM IN_LIST _RULE_UNSUPPORTED_PLATFORMS)
+    return()
+  endif()
+
+  if(NOT DEFINED _RULE_DRIVER)
+    message(SEND_ERROR "Need to specify the driver to use.")
+  endif()
+
+  set(_RUNNER_TARGET "iree-run-module")
+  iree_package_name(_PACKAGE_NAME)
+  set(_NAME "${_PACKAGE_NAME}_${_RULE_NAME}")
+
+  file(REAL_PATH "${_RULE_MODULE_SRC}" _SRC)
+  list(APPEND _RULE_RUNNER_ARGS "--module_file=${_SRC}")
+
+  if(_RULE_EXPECTED_OUTPUT)
+    # this may be a file or a literal output. In the latter case, the
+    # extension variable will be empty.
+    cmake_path(GET _RULE_EXPECTED_OUTPUT EXTENSION LAST_ONLY _OUTPUT_FILE_TYPE)
+    if(_OUTPUT_FILE_TYPE AND NOT
+       (_OUTPUT_FILE_TYPE STREQUAL ".txt" OR _OUTPUT_FILE_TYPE STREQUAL ".npy"))
+      message(SEND_ERROR
+        "Unsupported expected output file type: ${_RULE_EXPECTED_OUTPUT}"
+      )
```

nit: We don't enforce line length limits in CMake files, and I think this is easier to read without the extra line breaks

```suggestion
      message(SEND_ERROR "Unsupported expected output file type: ${_RULE_EXPECTED_OUTPUT}")
```

That keep the error handling easy to read past:
```cmake
if (foo AND bar)
  message(SEND_ERROR "...")
elseif(...)
  # continue 
endif()
```

---

Done.